common/scala/src/main/scala/whisk/core/database/ArtifactElasticSearchActivationStore.scala

```diff
@@ -0,0 +1,209 @@
+  logging: Logging,
+  override val httpFlow: Option[
+    Flow[(HttpRequest, Promise[HttpResponse]), (Try[HttpResponse], Promise[HttpResponse]), Any]] = None,
+  override val elasticSearchConfig: ElasticSearchActivationStoreConfig =
+    loadConfigOrThrow[ElasticSearchActivationStoreConfig](ConfigKeys.elasticSearchActivationStore))
+    extends ArtifactActivationStore(system, actorMaterializer, logging)
+    with ElasticSearchActivationRestClient {
+
+  implicit val m = actorMaterializer
+
+  val destinationDirectory: Path = Paths.get("logs")
+  val bufferSize = 100.MB
+  val perms = EnumSet.of(OWNER_READ, OWNER_WRITE, GROUP_READ, GROUP_WRITE, OTHERS_READ, OTHERS_WRITE)
+  protected val writeToFile: Sink[ByteString, _] = MergeHub
+    .source[ByteString]
+    .batchWeighted(bufferSize.toBytes, _.length, identity)(_ ++ _)
+    .to(RestartSink.withBackoff(minBackoff = 1.seconds, maxBackoff = 60.seconds, randomFactor = 0.2) { () =>
+      LogRotatorSink(() => {
+        val maxSize = bufferSize.toBytes
+        var bytesRead = maxSize
+        element =>
+          {
+            val size = element.size
+            if (bytesRead + size > maxSize) {
+              bytesRead = size
+              val logFilePath = destinationDirectory.resolve(s"userlogs-${Instant.now.toEpochMilli}.log")
+              logging.info(this, s"Rotating log file to '$logFilePath'")
+              try {
+                Files.createFile(logFilePath)
+                Files.setPosixFilePermissions(logFilePath, perms)
+              } catch {
+                case t: Throwable =>
+                  logging.error(this, s"Couldn't create activation record file '$t'")
+                  throw t
+              }
+              Some(logFilePath)
+            } else {
+              bytesRead += size
+              None
+            }
+          }
+      })
+    })
+    .run()
+
+  protected def extractRequiredHeaders(headers: Seq[HttpHeader]) =
+    headers.filter(h => elasticSearchConfig.requiredHeaders.contains(h.lowercaseName)).toList
+
+  protected def writeActivation(activation: WhiskActivation, context: UserContext) = {
+    val userIdField = Map("namespaceId" -> context.user.toJson)
+    val namespace = Map("namespace" -> activation.namespace.toJson)
+    val name = Map("name" -> activation.name.toJson)
+    val subject = Map("subject" -> activation.subject.toJson)
+    val activationId = Map("activationId" -> activation.activationId.toJson)
+    val start = Map("start" -> activation.start.toEpochMilli.toJson)
+    val end = Map("end" -> activation.end.toEpochMilli.toJson)
+    val cause = Map("cause" -> activation.cause.toJson)
+    val result = Map("result" -> activation.response.result.get.compactPrint.toJson)
+    val statusCode = Map("statusCode" -> activation.response.statusCode.toJson)
+    val logs = Map("logs" -> activation.logs.toJson)
+    val version = Map("version" -> activation.version.toJson)
+    val annotations = activation.annotations.toJsObject.fields
```

Comment:
Annotations fields need to be lifted to the root level of the JSON object. Maybe there is a cleaner way.